test/e2e/xfn_test.go

```diff
@@ -0,0 +1,344 @@
+			t.Fatalf("Cannot create secret %s: %v", secret.Name, err)
+		}
+		configMap := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "reg-ca",
+				Namespace: namespace,
+			},
+			Data: map[string]string{
+				"domain.crt": caPem,
+			},
+		}
+		if err := client.Create(ctx, configMap); err != nil {
+			t.Fatalf("Cannot create config %s: %v", configMap.Name, err)
+		}
+		return ctx
+	}
+}
+
+// CopyImagesToRegistry copies fn images to private registry
+func CopyImagesToRegistry() features.Func {
+	return func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
+		nodes := &corev1.NodeList{}
+		if err := config.Client().Resources().List(ctx, nodes); err != nil {
+			t.Fatal("cannot list nodes", err)
+		}
+		if len(nodes.Items) == 0 {
+			t.Fatalf("no nodes in the cluster")
+		}
+		var addr string
+		for _, a := range nodes.Items[0].Status.Addresses {
+			if a.Type == corev1.NodeInternalIP {
+				addr = a.Address
+				break
+			}
+		}
+		if addr == "" {
+			t.Fatalf("no nodes with private address")
+		}
+
+		srcRef, err := name.ParseReference("crossplane-e2e/fn-labelizer:latest")
+		if err != nil {
+			t.Fatal(err)
+		}
+		src, err := daemon.Image(srcRef)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		err = crane.Push(src, fmt.Sprintf("%s:32000/fn-labelizer:latest", addr), crane.Insecure)
+		if err != nil {
+			t.Fatal("copying image to registry not successful", err)
+		}
+		return ctx
+	}
+}
+
+// InstallDockerRegistry with custom TLS
+func InstallDockerRegistry() features.Func {
+	return funcs.AllOf(
+		funcs.AsFeaturesFunc(
+			funcs.HelmRepo(
+				helm.WithArgs("add"),
+				helm.WithArgs("twuni"),
+				helm.WithArgs("https://helm.twun.io"),
+			)),
+		funcs.AsFeaturesFunc(
+			funcs.HelmInstall(
+				helm.WithName("private"),
+				helm.WithNamespace("reg"),
+				helm.WithWait(),
+				helm.WithChart("twuni/docker-registry"),
+				helm.WithVersion("2.2.2"),
+				helm.WithArgs(
+					"--set service.type=NodePort",
+					"--set service.nodePort=32000",
+					"--set tlsSecretName=reg-cert",
+				),
+			)))
+}
+
+// CrossplaneDeployedWithFunctionsEnabled asserts that crossplane deployment with composition functions is enabled
+func CrossplaneDeployedWithFunctionsEnabled() features.Func {
+	return funcs.AllOf(
+		funcs.AsFeaturesFunc(funcs.HelmUpgrade(
+			HelmOptions(
+				helm.WithArgs(
+					"--set args={--debug,--enable-composition-functions}",
+					"--set xfn.enabled=true",
+					"--set xfn.args={--debug}",
+					"--set registryCaBundleConfig.name=reg-ca",
+					"--set registryCaBundleConfig.key=domain.crt",
+					"--set xfn.resources.requests.cpu=100m",
+					"--set xfn.resources.limits.cpu=100m",
+				),
+				helm.WithWait())...)),
+		funcs.ReadyToTestWithin(1*time.Minute, namespace),
+	)
+}
+
+// ProvideNopDeployed assets that provider-nop is deployed and healthy
+func ProvideNopDeployed() features.Func {
+	manifests := "test/e2e/manifests/xfnrunner/private-registry/pull/prerequisites"
```

Comment:
can you remove this function and unwrap it directly in the only place where it's called above? Please avoid hardcoding values such as manifests in functions as it could be misleading and lead to hard to debug errors.